src/alchemlyb/parsing/gomc.py

```diff
@@ -0,0 +1,211 @@
+    u_cols = [col for col in df.columns if any(single_u_col_match in col for single_u_col_match in u_col_match)]
+    u = None
+    if u_cols:
+        u = df[u_cols[0]]
+
+    u_k = dict()
+    cols = list()
+    for col in dH:
+        u_col = eval(col.split('->')[1][:-1])
+        # calculate reduced potential u_k = dH + pV + U
+        u_k[u_col] = beta * dH[col].values
+        if pv_cols:
+            u_k[u_col] += beta * pv.values
+        if u_cols:
+            u_k[u_col] += beta * u.values
+        cols.append(u_col)
+
+    u_k = pd.DataFrame(u_k, columns=cols,
+                       index=pd.Float64Index(times.values, name='time'))
+    # create columns for each lambda, indicating state each row sampled from
+    for i, l in enumerate(lambdas):
+        try:
+            u_k[l] = statevec[i]
+        except TypeError:
+            u_k[l] = statevec
+
+    # set up new multi-index
+    newind = ['time'] + lambdas
+    u_k = u_k.reset_index().set_index(newind)
+
+    u_k.name = 'u_nk'
+
+    return u_k
+
+
+def extract_dHdl(file, T):
+    """Return gradients `dH/dl` from a Hamiltonian differences free energy file.
+
+    Parameters
+    ----------
+    file : str
+        Path to free energy file to extract data from.
+    T : float
+        Temperature in Kelvin the simulations sampled.
+
+    Returns
+    -------
+    dH/dl : Series
+        dH/dl as a function of step for this lambda window.
+
+    """
+    beta = 1/(k_b * T)
+
+    state, lambdas, statevec = _extract_state(file)
+
+    # extract a DataFrame from free energy data
+    df = _extract_dataframe(file)
+
+    times = df[df.columns[0]]
+
+    # want to grab only dH/dl columns
+    dHcols = []
+    for l in lambdas:
+        dHcols.extend([col for col in df.columns if (l in col)])
+
+    dHdl = df[dHcols]
+
+    # make dimensionless
+    dHdl = beta * dHdl
+
+
+    dHdl = pd.DataFrame(dHdl.values, columns=lambdas,
+                        index=pd.Float64Index(times.values, name='time'))
+
+    # Need to modify the lambda name
+    cols = [l + "-lambda" for l in lambdas]
+    # create columns for each lambda, indicating state each row sampled from
+    for i, l in enumerate(cols):
+        try:
+            dHdl[l] = statevec[i]
+        except TypeError:
+            dHdl[l] = statevec
+
+    # set up new multi-index
+    newind = ['time'] + cols
+    dHdl= dHdl.reset_index().set_index(newind)
+
+    dHdl.name='dH/dl'
+
+    return dHdl
+
+
+def _extract_state(file):
+    """Extract information on state sampled, names of lambdas.
+
+    """
+    state = None
+    with anyopen(file, 'r') as f:
+        for line in f:
+            if ('#' in line) and ('State' in line):
+                state = int(line.split('State')[1].split(':')[0])
+                # GOMC always print these two field
+                lambdas = ['Coulomb', 'VDW']
+                statevec = eval(line.strip().split(' = ')[-1])
+                break
+
```

Comment:
I think this function parses the whole file. Is there a way that you can break once you have all necessary state information?

It's ok not to address this problem right now and leave it as an optimization for later (we have the same problem in other readers, too). But if it is the case that this can be optimized, add a comment saying so.

Reply:
In line 164, we break once we read all necessary state information. In GOMC output format, these information is printed in the first line. So, we never go through the whole file.